File: 0003_longest-substring-without-repeating-characters.py
# Trivial solution
# class Solution:
#     def lengthOfLongestSubstring(self, s: str) -> int:
#         N: int = len(s)

#         if N <= 1:
#             return N

#         max_len: int = 0
#         buff_str: str = ""
#         for i in range(N - 1):
#             buff_str = s[i]
#             for j in range(i + 1, N):
#                 char = s[j]
#                 if char in buff_str:
#                     max_len = max(max_len, len(buff_str))
#                     break
#                 else:
#                     buff_str += char
#             max_len = max(max_len, len(buff_str))

#         return max_len


# Trivial solution without len
# class Solution:
#     def lengthOfLongestSubstring(self, s: str) -> int:
#         N: int = len(s)
#         if N <= 1:
#             return N
#         max_len: int = 0
#         buff_str: str = ""
#         for i in range(N - 1):
#             buff_str = s[i]
#             for j in range(i + 1, N):
#                 char = s[j]
#                 if char in buff_str:
#                     max_len = max(max_len, j - i)
#                     break
#                 else:
#                     buff_str += char
#             else:
#                 max_len = max(max_len, N - i)

#         return max_len

# Solution using dicts instead of strings
# class Solution:
#     def lengthOfLongestSubstring(self, s: str) -> int:
#         N: int = len(s)
#         if N <= 1:
#             return N
#         max_len: int = 0
#         buff_dict: dict
#         for i in range(N - 1):
#             buff_dict = {}
#             buff_dict[s[i]] = None
#             for j in range(i + 1, N):
#                 char = s[j]
#                 if char in buff_dict:
#                     max_len = max(max_len, j - i)
#                     break
#                 else:
#                     buff_dict[char] = None
#             else:
#                 max_len = max(max_len, N - i)

#         return max_len


# Solution using a sliding window
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        N = len(s)
        max_len = 0
        idxs = [-1 for _ in range(128)]
        i = 0
        for j in range(N):
            if idxs[ord(s[j])] >= i:
                i = idxs[ord(s[j])] + 1
            idxs[ord(s[j])] = j
            max_len = max(max_len, j - i + 1)
        return max_len

File: 0003_longest-substring-without-repeating-characters.cpp
// -------------------------------------------------- //
//  Trivial not really optimal solution using str buffering
// --------------------------------------------------//
// include <algorithm>
// #include <iostream>
// #include <string>
// using namespace std;

// class Solution {
// public:
//   int lengthOfLongestSubstring(string s) {
//     const auto N = s.length();
//     if (N <= 1)
//       return N;
//     int max_len = 0;
//     for (int i = 0; i < N; i++) {
//       string buff = s.substr(i, 1);
//       int j = i + 1;
//       for (; j < N; j++) {
//         const auto c = s[j];
//         if (buff.find(c) != string::npos) {
//           max_len = max(max_len, j - i);
//           // cout << "break: " << max_len << endl;
//           break;
//         } else {
//           buff += s[j];
//           // cout << "non-break: " << buff << endl;
//         }
//       }
//       if (j == N) {
//         max_len = max(max_len, static_cast<int>(N - i));
//         // cout << "outer: " << max_len << endl;
//       }
//     }
//     return max_len;
//   }
// };

// -------------------------------------------------- //
//  Instead of a str buffer use an unordered_set
// --------------------------------------------------//
// #include <algorithm>
// #include <string>
// #include <unordered_set>
// using namespace std;

// class Solution {
// public:
//   int lengthOfLongestSubstring(string s) {
//     const auto N = s.length();
//     int maxLen = 0;
//     unordered_set<char> charTable;
//     int i = 0;
//     for (int j = 0; j < N; j++) {
//       if (charTable.count(s[j]) == 0) {
//         charTable.insert(s[j]);
//         maxLen = max(maxLen, j - i + 1);
//       } else {
//         while (charTable.count(s[j])) {
//           charTable.erase(s[i]);
//           i++;
//         }
//         charTable.insert(s[j]);
//       }
//     }

//     return maxLen;
//   }
// };

// -------------------------------------------------- //
//  Instead of traversing O(n log n) we create a sliding
//  window which is automatically shrinked on duplicates
//  => this results in linear complexity O(n)
// --------------------------------------------------//
#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    const auto N = s.length();
    vector<int> idxs(128, -1);
    int maxLen = 0;
    int i = 0;
    for (int j = 0; j < N; j++) {
      if (idxs[s[j]] >= i) {
        i = idxs[s[j]] + 1;
      }
      idxs[s[j]] = j;
      maxLen = max(maxLen, j - i + 1);
    }

    return maxLen;
  }
};
